### PR TITLE
Streaming

### PR DESCRIPTION
**Pull request by MinhNghiaD**
Hi, this is my second PR on the log streaming. Currently, I am using the Hook interface of logrus to receive file writing event. From this event, I registered the offset of the log entry in the file, this allow the stream to be able to seek and read the line instead of iterate the whole file. For each stream request, the server will create a log.Reader which is a subscriber to the Event Hook. It's receive a broadcast signal from the Hook and use it to trigger the read.

On the client side, I implemented a simple connection backoff mechanism to handle an interrupted stream. The maximum retry period is currently set to 1 minute.

Thank you all for reviewing and giving me your valuable advices.

### CODE REVIEW

File: api/client/client.go [hunk 1]
```diff
@@ -40,19 +42,117 @@ func (c *Client) Close() error {
...
+// Read reads the next log entry. Read implement connection backoff to handle connection error.
+// When a connection error occurs, the client will try to reset the connection and resume the stream instead of restart from the beginning.
+// The maximum retry period is 1 minute. After 1 minute, if the connection is still not fixed, Read returns the corresponding connection error.
+func (receiver *LogReceiver) Read() (line *proto.Log, err error) {
+	line = nil
+	const maxBackoff = time.Minute
+	currentBackoff := time.Second
+
+	// Attempt to read a log entry, with a retry period of maxBackoff
+	for line == nil && currentBackoff < maxBackoff {
```
Comment (russjones) [inline api/client/client.go:114]: What problem are you trying to solve here? Feels out of scope, if you get an error you don't need to back off and try again, just return the error.

File: cmd/worker_cli/worker_cli.go [hunk 2]
```diff
@@ -112,6 +118,32 @@ func queryJob(c *client.Client, jobID string) error {
...
 }
 
+// queryJob queries the status of a job specified by jobID
+func streamLog(c *client.Client, jobID string) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	receiver, err := c.GetLogReceiver(ctx, &proto.Job{Id: jobID})
+	var entry *proto.Log = nil
+
+	for err == nil {
```
Comment (russjones) [inline cmd/worker_cli/worker_cli.go:128]: Minor, you don't need to make this change, but I try and follow common patterns to make reviewing code like this easier for others.

```go
for {
   in, err := stream.Recv()
   if err == io.EOF {
     return nil
   }
   if err != nil {
     return err
   }
   // do some work
}
```

File: api/worker/proto/worker_service.proto [hunk 3]
```diff
@@ -40,9 +40,18 @@ message StopRequest {
...
+message StreamRequest {
+    // The Job to be stopped.
+    Job job = 1;
+    // the nb of sequence of the starting point of the stream
+    int32 startPoint = 2;
+}
+
 message Log {
+    // Nb of sequence of the current entry in the stream
+    int32 nbSequence = 1;
```
Comment (russjones) [inline api/worker/proto/worker_service.proto:52]: I don't beleve this was documented in the design document, why not just start the stream from the beginning?

Comment (MinhNghiaD) [reply api/worker/proto/worker_service.proto:52]: Hi, I think that I have mentioned it in the streaming design section. The connection backoff is for long-live streaming

File: cmd/worker_cli/worker_cli.go (hunk 2 above)
Comment (MinhNghiaD) [reply cmd/worker_cli/worker_cli.go:128]: Thank you, I will fix it

File: api/client/client.go (hunk 1 above)
Comment (MinhNghiaD) [reply api/client/client.go:114]: I was attempting to implement long live streaming connection backoff mechanism, as I have mentioned in the streaming section of the design document. If you think that it is unnecessary, I will delete it
